AnimatedComponents/src/components/timer/Timer.tsx

```diff
@@ -0,0 +1,191 @@
+import Animated, {
+  Easing,
+  cancelAnimation,
+  useAnimatedProps,
+  useAnimatedStyle,
+  useSharedValue,
+  withTiming,
+} from 'react-native-reanimated'
+import Svg, { Circle as SvgCircle } from 'react-native-svg'
+
+import { TimerPreset } from '@constants'
+
+import styles from './timer-styles'
+
+const AnimatedSvgCircle = Animated.createAnimatedComponent(SvgCircle)
+
+interface ITimerProps {
+  /**buttonStyles: is an optional prop to style the buttons of the timer */
+  buttonStyles?: { container?: ViewStyle; text?: TextStyle }
+  /**circularTimerStrokeColor: is an optional prop that indicates the stroke color of circular timer */
+  circularTimerStrokeColor?: string
+  /**controls: is an optional prop that indicates whether control buttons are visible or not */
+  controls?: boolean
+  /**linearTimerColor: is an optional prop that indicates the color of linear timer */
+  linearTimerColor?: string
+  /**timerType: is a required prop that indicates the type of timer, circular or linear */
+  timerType: TimerPreset
+  /**totalDurationInSeconds: is a required prop that indicates the total number of seconds of timer */
+  totalDurationInSeconds: number
+}
+
+const Timer = (props: ITimerProps) => {
+  const {
+    buttonStyles,
+    circularTimerStrokeColor = 'black',
+    controls = false,
+    linearTimerColor = 'black',
+    timerType,
+    totalDurationInSeconds,
+  } = props
+
+  const [timeLeft, setTimeLeft] = useState(totalDurationInSeconds)
+  const [isRunning, setIsRunning] = useState(false)
+  const [isPaused, setIsPaused] = useState(false)
+
+  const progress = useSharedValue(1)
+
+  const handleControl = useCallback(
+    (action: 'start' | 'pause' | 'resume') => {
+      switch (action) {
+        case 'start':
+          setIsRunning(true)
+          setIsPaused(false)
+          startAnimation(timeLeft * 1000)
+          break
+        case 'pause':
+          setIsRunning(false)
+          setIsPaused(true)
+          cancelAnimation(progress)
+          break
+        case 'resume':
+          setIsRunning(true)
+          setIsPaused(false)
+          startAnimation(timeLeft * 1000)
+          break
+      }
+    },
+    [timeLeft, progress],
+  )
+
+  const startAnimation = (duration: number) => {
+    progress.value = withTiming(0, { duration, easing: Easing.linear })
+  }
+
+  const handleReset = useCallback(() => {
+    setIsRunning(false)
+    setIsPaused(false)
+    setTimeLeft(totalDurationInSeconds)
+    progress.value = 1
+  }, [totalDurationInSeconds, progress])
+
+  const animatedCircleProps = useAnimatedProps(() => ({
+    strokeDashoffset: (1 - progress.value) * 2 * Math.PI * 45,
+  }))
+
+  const animatedLinearStyle = useAnimatedStyle(() => ({
+    width: `${progress.value * 100}%`,
+    backgroundColor: linearTimerColor,
+  }))
+
+  useEffect(() => {
+    if (isRunning) {
+      const interval = setInterval(() => {
+        setTimeLeft(prev => (prev > 0 ? prev - 1 : 0))
```

Comment:
we try avoiding states in animation. as well as timeout's/ interval

we can do instead
```  useEffect(() => {
    reanimatedTimer.value = withTiming(0, { duration: duration * 1000, easing: ReanimatedEasing.linear })

    return () => {
      reanimatedTimer.value = 1
    }
  }, [])


  ```